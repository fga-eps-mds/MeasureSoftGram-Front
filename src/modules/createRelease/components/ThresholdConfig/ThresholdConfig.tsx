import React, { useState, useCallback } from 'react';
import { Box, Grid, Typography, Link } from '@mui/material';

import { Characteristic, Measure, Subcharacteristic } from '@customTypes/preConfig';
import CheckboxButton from '@components/CheckboxButton/CheckboxButton';
import undelineRemover from '@utils/undelineRemover';

import { useCreateReleaseContext } from '@modules/createRelease/context/useCreateRelease';
import getThresholdInfo from '@modules/createRelease/utils/getThresholdInfo';
import { componentIterator } from '../../utils/componentIterator';

import PreConfigTabs from '../PreConfigTabs';
import ThresholdSlider from '../ThresholdSlider';

interface ThresholdConfigProrps {
  data: Characteristic[];
  setCheckboxValues: Function;
  checkboxValues: string[];
  tabs: string[];
}

type limiterType = { tabName?: string; data: { key: string; weight: number } };

const ThresholdConfig = ({ data, checkboxValues, setCheckboxValues, tabs }: ThresholdConfigProrps) => {
  const [tabValue, setTabValue] = useState<string>(tabs[0]);
  const [limiters, setLimiters] = useState<limiterType[]>([]);
  const { setCurrentConfig } = useCreateReleaseContext();

  const keyGetter = (objectArray: limiterType[]) => objectArray.map((object) => object.data.key);

  const checkboxValue = useCallback(
    (key: string) => {
      let namesArray = [...checkboxValues];

      if (checkboxValues.includes(key)) {
        namesArray = namesArray.filter((name) => name !== key);
      } else namesArray.push(key);

      setCheckboxValues(namesArray);
    },
    [checkboxValues, setCheckboxValues]
  );

  const onChange = (min: number, max: number, key: string) => {
    const thresholdInfo = getThresholdInfo(key);

    const isOutOfRange = (thresholdInfo?.range[0] ?? 0) > min || (thresholdInfo?.range[1] ?? Infinity) < max;

    if (isOutOfRange || max < min) {
      return;
    }

    const mapMeasureValue = (measureValue: Measure): Measure => {
      if (measureValue.key !== key) {
        return measureValue;
      }

      const { minFixed, maxFixed } = thresholdInfo ?? {};
      const newMin = minFixed ? measureValue.min_threshold : min;
      const newMax = maxFixed ? measureValue.max_threshold : max;

      return {
        ...measureValue,
        min_threshold: newMin,
        max_threshold: newMax
      };
    };

    const tempData = data.map((charValue) => ({
      ...charValue,
      subcharacteristics: charValue.subcharacteristics.map((subCharValue) => ({
        ...subCharValue,
        measures: subCharValue.measures.map(mapMeasureValue)
      }))
    }));

    setCurrentConfig(tempData);
  };

  const checkBoxCallback = (
    value: Measure | Characteristic | Subcharacteristic,
    previousValue: Characteristic | Subcharacteristic
  ) => {
    if (previousValue?.key === tabValue || !previousValue) {
      const isChecked = checkboxValues.includes(value.key);
      return (
        <Grid item key={value.key}>
          <CheckboxButton
            label={undelineRemover(value.key)}
            checked={isChecked}
            backgroundColor="#113d4c"
            colorHover="#113d4c"
            style={{ marginRight: '8px', marginBottom: '10px' }}
            onClick={() => {
              checkboxValue(value.key);
              const index = keyGetter(limiters).indexOf(value.key);
              if (!(index < 0)) {
                limiters.splice(index, 1);
                const newLimiterValue = limiters;
                setLimiters(newLimiterValue);
              }
            }}
          />
        </Grid>
      );
    }
  };

  const renderCheckBoxes = () => (
    <Grid container marginBottom="15px" columns={4}>
      {componentIterator.measure(data, checkBoxCallback)}
    </Grid>
  );

  const renderSliderCallback = (value: Measure, previousValue: Subcharacteristic) => {
    if (checkboxValues.includes(value.key) && (!previousValue || previousValue.key === tabValue)) {
      const tabName = previousValue?.key;
      const thresholdInfo = getThresholdInfo(value.key);
      const minThreshold = value.min_threshold ?? thresholdInfo?.range[0] ?? 0
      const maxThreshold = value.max_threshold ?? thresholdInfo?.range[1] ?? 100

      // if don't find the key in the limiters array, add it
      if (keyGetter(limiters).indexOf(value.key) === -1) {
        const sliderValue: limiterType = { tabName, data: { key: value.key, weight: value.weight } };
        const newLimiterValue: limiterType[] = [...limiters, sliderValue];
        setLimiters(newLimiterValue);
      }
      return (
        <ThresholdSlider
          key={value.key}
          label={value.key}
          min={minThreshold}
          max={maxThreshold}
          onChange={onChange}
        />
      );
    }
  };

  const renderSliders = () => <Box marginBottom="30px">{componentIterator.measure(data, renderSliderCallback)}</Box>;

  const setTab = (_e: any, newValue: string) => {
    setTabValue(newValue);
  };

  const renderTabs = () => {
    if (tabs)
      return <PreConfigTabs style={{ marginBottom: '24px' }} onChange={setTab} value={tabValue} tabsValues={tabs} />;
  };

  return (
    <Box display="flex" flexDirection="column" maxWidth="600px">
      <Typography variant="h4" sx={{ marginBottom: '16px', color: '#33568E', fontWeight: '500' }}>
        Definir os valores de referência mínimo e máximo
      </Typography>
      <Typography variant="body1" sx={{ marginBottom: '16px' }}>
        Os valores de referência afetam como algumas medidas são calculadas. Aluns deles não podem ser modificados. Para
        mais detalhes veja a documentação disponível em{' '}
        <Link
          href="https://fga-eps-mds.github.io/2023-1-MeasureSoftGram-Doc/manual-de-uso/cli/#2311-thresholds"
          underline="hover"
          target="_blank"
        >
          (link)
        </Link>
      </Typography>
      {renderTabs()}
      {renderCheckBoxes()}
      <Grid container columns={2}>
        <Grid item xs={1}>
          <Typography variant="body1" textAlign="start">
            Min
          </Typography>
        </Grid>
        <Grid item xs={1}>
          <Typography variant="body1" textAlign="end">
            Max
          </Typography>
        </Grid>
      </Grid>
      {renderSliders()}
    </Box>
  );
};

export default ThresholdConfig;
