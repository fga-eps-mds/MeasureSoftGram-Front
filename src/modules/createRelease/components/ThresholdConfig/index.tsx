import ThresholdConfig from "./ThresholdConfig";

export default ThresholdConfig;
