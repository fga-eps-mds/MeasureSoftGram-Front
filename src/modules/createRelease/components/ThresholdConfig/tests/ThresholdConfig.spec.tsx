import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ThresholdConfig from '../ThresholdConfig';

interface TestingComponentProps {
  onChange: (min: number, max: number, key: string) => void;
  label: string;
}

const TestingComponent = ({ onChange, label }: TestingComponentProps) => (
  <div>
    <input
      type='number'
      data-testid='threshold-slider'
      onChange={(e) => onChange(Number(e.target.value), Number(e.target.value), label)}
    />
  </div>);

jest.mock('@modules/createRelease/components/ThresholdSlider/ThresholdSlider', () => TestingComponent);

jest.mock('@modules/createRelease/context/useCreateRelease', () => ({
  useCreateReleaseContext: () => ({
    setCurrentConfig: jest.fn(),
  })
}));

describe('<ThresholdConfig />', () => {
  const data = [
    {
      key: 'characteristic1',
      weight: 100,
      subcharacteristics: [
        {
          key: 'subcharacteristic1',
          weight: 70,
          measures: [
            { key: 'measure1', min_threshold: 10, max_threshold: 20, weight: 1 },
            { key: 'measure2', min_threshold: 30, max_threshold: 40, weight: 2 },
          ],
        },
        {
          key: 'subcharacteristic2',
          weight: 30,
          measures: [
            { key: 'measure3', min_threshold: 50, max_threshold: 60, weight: 3 },
            { key: 'measure4', min_threshold: 70, max_threshold: 80, weight: 4 },
          ]
        }
      ],
    },
  ];

  describe("Snapshot", () => {
    it('Deve chamar a função setCheckboxValues ao clicar na checkbox', () => {
      const setCheckboxValues = jest.fn();

      const tree = render(
        <ThresholdConfig
          data={data}
          checkboxValues={['measure1']}
          setCheckboxValues={setCheckboxValues}
          tabs={['subcharacteristic1']}
        />
      );

      expect(tree).toMatchSnapshot();
    })
  });

  it('Deve renderizar as checkboxes corretamente', () => {
    const setCheckboxValues = jest.fn();

    const { getByText } = render(
      <ThresholdConfig
        data={data}
        checkboxValues={['measure1']}
        setCheckboxValues={setCheckboxValues}
        tabs={['subcharacteristic1']}
      />
    );
    const checkboxMeasure1 = getByText('measure1');
    const checkboxMeasure2 = getByText('measure2');

    expect(checkboxMeasure1).toBeDefined();
    expect(checkboxMeasure2).toBeDefined();
    fireEvent.click(checkboxMeasure1);
    fireEvent.click(checkboxMeasure2);
    expect(setCheckboxValues).toBeCalled();
  });

  it('Deve chamar a função onChange ao alterar os sliders', () => {
    const { getByTestId } = render(
      <ThresholdConfig
        data={data}
        checkboxValues={['measure1']}
        setCheckboxValues={jest.fn()}
        tabs={['subcharacteristic1']}
      />
    );

    const sliderInput = getByTestId('threshold-slider');
    expect(sliderInput).toBeDefined();
    fireEvent.change(sliderInput, { target: { value: 50 } })
    fireEvent.change(sliderInput, { target: { value: -50 } })
  });

  it('Deve clicar na tab', () => {
    const { getByText } = render(
      <ThresholdConfig
        data={data}
        checkboxValues={['measure1']}
        setCheckboxValues={jest.fn()}
        tabs={['subcharacteristic1', 'subcharacteristic2']}
      />
    );

    const subcharacteristicTab = getByText('subcharacteristic2');
    expect(subcharacteristicTab).toBeDefined();
    fireEvent.click(subcharacteristicTab);
  });

});
