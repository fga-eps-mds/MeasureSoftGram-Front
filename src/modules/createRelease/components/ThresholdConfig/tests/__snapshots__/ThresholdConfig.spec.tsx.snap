// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThresholdConfig /> Snapshot Deve chamar a função setCheckboxValues ao clicar na checkbox 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-1i9whfo"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-pxw7rc-MuiTypography-root"
        >
          Definir os valores de referência mínimo e máximo
        </h4>
        <p
          class="MuiTypography-root MuiTypography-body1 css-z2eky3-MuiTypography-root"
        >
          Os valores de referência afetam como algumas medidas são calculadas. Aluns deles não podem ser modificados. Para mais detalhes veja a documentação disponível em Link
        </p>
        <div
          class="MuiTabs-root css-1ujnqem-MuiTabs-root"
          style="margin-bottom: 24px;"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
            style="overflow: hidden; margin-bottom: 0px;"
          >
            <div
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                role="tab"
                tabindex="0"
                type="button"
              >
                subcharacteristic1
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-s3g6yz-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              style="margin-right: 8px; margin-bottom: 10px;"
            >
              <input
                checked=""
                name="measure1"
                style="position: absolute; opacity: 0;"
                type="checkbox"
              />
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1cfee4v-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                measure1
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <div
              style="margin-right: 8px; margin-bottom: 10px;"
            >
              <input
                name="measure2"
                style="position: absolute; opacity: 0;"
                type="checkbox"
              />
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-cjtmh9-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                measure2
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1v2d8t3-MuiTypography-root"
            >
              Min
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-f8hnn-MuiTypography-root"
            >
              Max
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-1xn3edc"
        >
          <div>
            <input
              data-testid="threshold-slider"
              type="number"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-1i9whfo"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-pxw7rc-MuiTypography-root"
      >
        Definir os valores de referência mínimo e máximo
      </h4>
      <p
        class="MuiTypography-root MuiTypography-body1 css-z2eky3-MuiTypography-root"
      >
        Os valores de referência afetam como algumas medidas são calculadas. Aluns deles não podem ser modificados. Para mais detalhes veja a documentação disponível em Link
      </p>
      <div
        class="MuiTabs-root css-1ujnqem-MuiTabs-root"
        style="margin-bottom: 24px;"
      >
        <div
          class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
          style="overflow: hidden; margin-bottom: 0px;"
        >
          <div
            class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
            role="tablist"
          >
            <button
              aria-selected="true"
              class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
              role="tab"
              tabindex="0"
              type="button"
            >
              subcharacteristic1
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <span
            class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
            style="left: 0px; width: 0px;"
          />
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-s3g6yz-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <div
            style="margin-right: 8px; margin-bottom: 10px;"
          >
            <input
              checked=""
              name="measure1"
              style="position: absolute; opacity: 0;"
              type="checkbox"
            />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1cfee4v-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              measure1
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <div
            style="margin-right: 8px; margin-bottom: 10px;"
          >
            <input
              name="measure2"
              style="position: absolute; opacity: 0;"
              type="checkbox"
            />
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-cjtmh9-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              measure2
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1v2d8t3-MuiTypography-root"
          >
            Min
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1 css-1osj8n2-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-f8hnn-MuiTypography-root"
          >
            Max
          </p>
        </div>
      </div>
      <div
        class="MuiBox-root css-1xn3edc"
      >
        <div>
          <input
            data-testid="threshold-slider"
            type="number"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
